src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,299 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
-			break;
-		default:
-			LOG_TARGET_ERROR(target, "Unsupported size: %d", size);
-			result = ERROR_FAIL;
-			goto error;
+	for (target_addr_t address = start_address;
+			address - start_address < distance;
+			address += size, buffer += size) {
+		uint64_t value = get_from_buffer_and_log_write(address, buffer, size);
+		if (size > 4)
+			riscv_batch_add_dm_write(batch, DM_DATA1, value >> 32, false);
+		riscv_batch_add_dm_write(batch, DM_DATA0, value, false);
 	}
+	return start_address + distance;
+}
 
-	if (riscv_enable_virtual && has_sufficient_progbuf(target, 5) && get_field(mstatus, MSTATUS_MPRV))
-		riscv_program_csrrci(&program, GDB_REGNO_ZERO,  CSR_DCSR_MPRVEN, GDB_REGNO_DCSR);
-	riscv_program_addi(&program, GDB_REGNO_S0, GDB_REGNO_S0, size);
+/**
+ * This function runs the batch of writes and updates address_p with the
+ * address of the next write.
+ */
+static int write_memory_progbuf_inner_run_batch(struct target *target, struct riscv_batch *batch,
+		target_addr_t *address_p, target_addr_t end_address, uint32_t size,
+		const uint8_t *buffer)
+{
+	if (batch_run(target, batch) != ERROR_OK)
+		return ERROR_FAIL;
 
-	result = riscv_program_ebreak(&program);
-	if (result != ERROR_OK)
-		goto error;
-	riscv_program_write(&program);
+	/* Note that if the scan resulted in a Busy DMI response, it
+	 * is this read to abstractcs that will cause the dmi_busy_delay
+	 * to be incremented if necessary. */
+	uint32_t abstractcs;
 
-	riscv_addr_t cur_addr = address;
-	riscv_addr_t distance = (riscv_addr_t)count * size;
-	bool setup_needed = true;
-	LOG_TARGET_DEBUG(target, "Writing until final address 0x%016" PRIx64, cur_addr + distance);
-	while (cur_addr - address < distance) {
-		LOG_TARGET_DEBUG(target, "Transferring burst starting at address 0x%016" PRIx64,
-				cur_addr);
+	if (wait_for_idle(target, &abstractcs) != ERROR_OK)
+		return ERROR_FAIL;
+	RISCV013_INFO(info);
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
 
-		struct riscv_batch *batch = riscv_batch_alloc(
-				target,
-				RISCV_BATCH_ALLOC_SIZE,
-				info->dmi_busy_delay + info->ac_busy_delay);
-		if (!batch)
-			goto error;
+	bool dmi_busy_encountered = riscv_batch_dmi_busy_encountered(batch);
+	if (info->cmderr == CMDERR_NONE && !dmi_busy_encountered) {
+		LOG_TARGET_DEBUG(target, "Successful memory write M[0x%" TARGET_PRIxADDR
+				", 0x%" TARGET_PRIxADDR ")", *address_p, end_address);
+		*address_p = end_address;
+		return ERROR_OK;
+	} else if (info->cmderr == CMDERR_BUSY || dmi_busy_encountered) {
+		if (info->cmderr == CMDERR_BUSY)
+			LOG_TARGET_DEBUG(target, "Memory write resulted in abstract command busy response.");
+		if (dmi_busy_encountered)
+			LOG_TARGET_DEBUG(target, "Memory write resulted in DMI busy response.");
+		/* TODO: If dmi busy is encountered, the address of the last
+		 * successful write can be deduced by analysing the batch.
+		 */
+		return write_memory_progbuf_inner_on_busy(target, address_p, size, buffer);
```

Comment:
This can be given a more descriptive name:

```suggestion
		return write_memory_progbuf_inner_handle_busy(target, address_p, size, buffer);
```

or 

```suggestion
		return write_memory_progbuf_inner_get_failed_addr(target, address_p, size, buffer);
```

Reply:
Addressed.